drivers/i2c/i2c_ite_enhance.c

```diff
@@ -0,0 +1,771 @@
+		break;
+	case DT_REG_ADDR(DT_NODELABEL(i2c5)):
+		offset = CGC_OFFSET_SMBF;
+		break;
+	}
+
+	volatile uint8_t *reg = (volatile uint8_t *)
+				(IT8XXX2_ECPM_BASE + (offset >> 8));
+	uint8_t reg_mask = offset & 0xff;
+	*reg &= ~reg_mask;
+
+	/* Enable I2C function */
+	/* Software reset */
+	IT8XXX2_I2C_DHTR(base) |= IT8XXX2_I2C_SOFT_RST;
+	IT8XXX2_I2C_DHTR(base) &= ~IT8XXX2_I2C_SOFT_RST;
+	/* reset i2c port */
+	i2c_reset(dev);
+	/* bit1, Module enable */
+	IT8XXX2_I2C_CTR1(base) = 0;
+
+	/* Set clock frequency for I2C ports */
+	bitrate_cfg = i2c_map_dt_bitrate(config->bitrate);
+	error = i2c_enhance_configure(dev, I2C_MODE_MASTER | bitrate_cfg);
+	data->i2ccs = I2C_CH_NORMAL;
+
+	if (error) {
+		LOG_ERR("i2c: failure initializing");
+		return error;
+	}
+
+	/* The pin is set to I2C alternate function of SCL */
+	pinmux_pin_set(config->alts_list[SCL].pinctrls,
+		       config->alts_list[SCL].pin,
+		       config->alts_list[SCL].alt_fun);
+	/* The pin is set to I2C alternate function of SDA */
+	pinmux_pin_set(config->alts_list[SDA].pinctrls,
+		       config->alts_list[SDA].pin,
+		       config->alts_list[SDA].alt_fun);
+
+	return 0;
+}
+
+static int i2c_enhance_recover_bus(const struct device *dev)
+{
+	const struct i2c_enhance_config *config = dev->config;
+	int i;
+
+	/* Set SCL of I2C as GPIO pin */
+	pinmux_pin_input_enable(config->alts_list[SCL].pinctrls,
+				config->alts_list[SCL].pin,
+				PINMUX_OUTPUT_ENABLED);
+	/* Set SDA of I2C as GPIO pin */
+	pinmux_pin_input_enable(config->alts_list[SDA].pinctrls,
+				config->alts_list[SDA].pin,
+				PINMUX_OUTPUT_ENABLED);
+
+	/* Pull SCL and SDA pin to high */
+	gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 1);
+	gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 1);
+	k_msleep(1);
+
+	/* Start condition */
+	gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 0);
+	k_msleep(1);
+	gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 0);
+	k_msleep(1);
+
+	/* 9 cycles of SCL with SDA held high */
+	for (i = 0; i < 9; i++) {
+		/* SDA */
+		gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 1);
+		/* SCL */
+		gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 1);
+		k_msleep(1);
+		/* SCL */
+		gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 0);
+		k_msleep(1);
+	}
+	/* SDA */
+	gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 0);
+	k_msleep(1);
+
+	/* Stop condition */
+	gpio_pin_set(config->gpio_dev, config->alts_list[SCL].pin, 1);
+	k_msleep(1);
+	gpio_pin_set(config->gpio_dev, config->alts_list[SDA].pin, 1);
+	k_msleep(1);
+
+	/* Set GPIO back to I2C alternate function of SCL */
+	pinmux_pin_set(config->alts_list[SCL].pinctrls,
+		       config->alts_list[SCL].pin,
+		       config->alts_list[SCL].alt_fun);
+	/* Set GPIO back to I2C alternate function of SDA */
+	pinmux_pin_set(config->alts_list[SDA].pinctrls,
+		       config->alts_list[SDA].pin,
+		       config->alts_list[SDA].alt_fun);
```

Comment:
can this procedure be documented? why sleeps of 1ms?

Reply:
Add comment.